### PR TITLE
Design for serving models with different backends

### PR DESCRIPTION
**Pull request by alimaredia**
This commit introduces a new design document for the [`ilab model
serve`] command, detailing its
functionality to serve models using different backends, specifically
`llama-cpp` and `vllm`. It outlines the command structure, including
backend-specific flags and arguments, and proposes a testing strategy
for new engine integrations.

Co-authored-by: Sébastien Han <seb@redhat.com>
Signed-off-by: Ali Maredia <amaredia@redhat.com>

### CODE REVIEW

File: docs/cli/ilab-model-serve.md
Comment (leseb) [inline docs/cli/ilab-model-serve.md:17]: As mentioned earlier, this feels unnatural: `ilab model serve llama-cpp`, I would imagine something like `ilab model serve --backend llama-cpp` instead. If additional backends or more complex configurations are added, the command structure might become less clear. Using `--backend` is immediately clear that `--backend` is an option, making the command more self-explanatory.
When designing the CLI, I like to remind myself of the following:

```bash
program action --flag <option>
```

Here llama-cpp is not an action, it's an option. That's how I see things, also understand that people might think otherwise :)

File: docs/cli/ilab-model-serve.md
Comment (leseb) [inline docs/cli/ilab-model-serve.md:70]: A point I've raised before, I'm not sure if it makes sense to prefix everything with `--llama-cpp` if we know we are using the llama-cpp backend.

File: docs/cli/ilab-model-serve.md
Comment (leseb) [inline docs/cli/ilab-model-serve.md:74]: I don't think we should rely on the extension but have a function/script to read the image and tell us the format instead.

File: docs/cli/ilab-model-serve.md
Comment (leseb) [inline docs/cli/ilab-model-serve.md:86]: The only issue with this approach is how can we easily translate a given arg to an internal value? With a flag that is very easy to do. The downside with the flag is that we keep playing catching with every flag vllm produces...

**Review by leseb — COMMENTED** [review]
Thanks for working on this proposal! I envision an ideal scenario where we can determine the appropriate backend solely based on the model input (directory or file). I'm questioning whether this is something users really need to worry about. If we can alleviate this hassle, I believe we should.

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:86]: I don't think I'm being very clear here, the intention of `--vllm-args` is to not play catch up with every flag vllm has, but instead is intended to be a catch-all flag for all of it's arguments, which is something I think you'd support.

For example the community deams the [vllm arguments](https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#command-line-arguments-for-the-server) `--chat-template` and `--download-dir` to be ones that we use often and want to have flags for the subcommand. 

Those would get prefixed with `--vllm-` and become `--vllm-chat-template` and `--vllm-download-dir`. 

If a user wanted to specify more "obscure" arguments to vllm like the `--kv-cache-dtype` and `--block-size` when serving vllm, an example of a command would be:

`ilab model serve vllm --vllm-download-dir <PATH> --vllm-args "--kv-cache-dtype auto --block-size 16"`.

The value of `--vllm-args` is just getting passed straight into the vllm upon startup. The main function of the openai compatible server take and processes all arguments on start up.

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:74]: I can make this statement more general, the point of the bullet is that the command auto detects that it can't serve the format specified and throws an error with feedback for the user.

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:70]: The reason I made it so specific is that I'm scarred by the flags for training that exist in the codebase as of 6/7/24. It's not clear which training flags belong with the mac implementation of training vs the linux implementation and on top of that the flags are very similar.

An example of how the flags could be mixed up. Current ilab serve with llama-cpp has a flag for `max-ctx-size`. `vllm` has a flag for model context length called `--max-model-len` (https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#command-line-arguments-for-the-server). 

We could do the following possibilities:
1. Have a flag for `--max-model-len` and `--max-ctx-size` similar to how we have it in current `ilab train`. I think this option is hardest on the user.

2. Have one flag for context size called `--max-ctx-size` whose value is filtered down to each backend. An issue then becomes what do you do with flags that one backend has that one doesn't. Also you've now added this flag into the set of flags that both backends can use. 

In my current design I propose we make backend agnostic flags those flags have nothing to do with the configuration of the backend itself and more for the configuration of `ilab` and what the backends are operating on (like the model path).

3. Prefix all flags related to a backend with the backend name. I think this is the most clear for the users with the downside that it's more verbose.

@cdoern what do you think about how the flags for training might look. Training will have a similar set of possible subcommands with numerous options.

Comment (n1hility) [reply docs/cli/ilab-model-serve.md:70]: > In my current design I propose we make backend agnostic flags those flags have nothing to do with the configuration of the backend itself and more for the configuration of ilab and what the backends are operating on (like the model path).
 +1 this is the right approach to create and use generic flags yet still have a mechanism for provider specific flags.

File: docs/cli/ilab-model-serve.md
Comment (n1hility) [reply docs/cli/ilab-model-serve.md:74]: One other point to consider is that it's likely we may have multiple backends that support overlapping formats. To address that we could have a config flag to specify the intended default backend. Although that can be a later design enhancement. I only mention it to show the existing design here is compatible with expansion

File: docs/cli/ilab-model-serve.md
Comment (n1hility) [reply docs/cli/ilab-model-serve.md:86]: Another example of why this is necessary is that a user also needs to sometimes use hidden behavioral flags that do not have a long lifespan to work around a problem in the backend. Yet another case would be cuda flags that vllm needs to pass through

**Review by n1hility — APPROVED** [review]

File: docs/cli/ilab-model-serve.md
Comment (n1hility) [reply docs/cli/ilab-model-serve.md:17]: IMO either option works well no strong opinion from me :)

File: docs/cli/ilab-model-serve.md
Comment (mairin) [inline docs/cli/ilab-model-serve.md:58]: Yes! Love this mechanism. Gives us a way to be opinionated - provide a good default - but also flexible enough that the user who has their own opinions can express them!

File: docs/cli/ilab-model-serve.md
Comment (mairin) [inline docs/cli/ilab-model-serve.md:65]: is there an inconsistency here? Up on line 51 you're calling it `--model` - here on 65 you're calling it `--model-path` ?

File: docs/cli/ilab-model-serve.md
Comment (mairin) [reply docs/cli/ilab-model-serve.md:74]: @n1hility Agreed. Great idea.

File: docs/cli/ilab-model-serve.md
Comment (mairin) [inline docs/cli/ilab-model-serve.md:84]: Instead of error out, I'd suggest it suggest the correct backend. E.g., 


`You specified vllm but you've provided a gguf file which is not supported by vllm. Would you like to run the gguf file using llamacpp?`

File: docs/cli/ilab-model-serve.md
Comment (mairin) [reply docs/cli/ilab-model-serve.md:86]: I like the catchall approach. It's one we used to handle arguments to plugins we don't control in the ChRIS project. It has worked well.

**Review by mairin — CHANGES_REQUESTED** [review]

File: docs/cli/ilab-model-serve.md
Comment (leseb) [reply docs/cli/ilab-model-serve.md:86]: Ok now it's clearer, I'm ok with the `--vllm-args` value being passed as is straight upon vllm startup.

File: docs/cli/ilab-model-serve.md
Comment (leseb) [reply docs/cli/ilab-model-serve.md:17]: @mairin I don't think the second bullet can apply given that we are going the route of generic flags and not re-using backend-specific flags. This is described here: https://github.com/instructlab/dev-docs/pull/81/files#r1631408310

Other arguments from me on why I don't like the positional argument in this scenario:

* clarity: I know the value that I'm passing to `--backend` is a backend, when I pass `llama-cpp`, I might have no clue what it is
* stability: once we introduce `ilab model serve llama-cpp` we are blocked from adding any other things than a backend. Using `--backend` removes that problem. What if one day we want to add another verb/noun to `ilab model serve <whatever>`.
* consistency: it's a common CLI design practice to use flags to specify options

File: docs/cli/ilab-model-serve.md
Comment (leseb) [reply docs/cli/ilab-model-serve.md:70]: I'm open to building generic flags, but it feels like we're creating a third layer of abstraction, which might introduce some ambiguity. Users will have to juggle raw flags from llama-cpp, as well as flags from ilab, which could complicate their mental model when writing commands. However, I do like the idea of re-using sensible flags and handling the backend mapping internally.

In the end, we could adopt two strategies:

* Backend-Agnostic Flags: Select the most common flags that make sense across multiple backends.
* Backend-Specific Flags: Use --vllm-args (or similar) for backend-specific configurations.

There might still be issues, such as someone passing ilab model serve --backend llama-cpp --llama-cpp-max-ctx-size 1 --llama-cpp-args "llama-cpp-max-ctx-size 2". However, we can address this through proper validation and error messaging.

File: docs/cli/ilab-model-serve.md
Comment (russellb) [reply docs/cli/ilab-model-serve.md:17]: One point that helps me decide these choices --

For llamacpp vs vllm, are the rest of the --flags identical? If not I'd prefer separate commands vs a --backend flag. I would prefer only doing something like `--backend` if the rest of the flags apply equally to both (and any future backend)

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:84]: I think you're suggesting that the command would steer the user to use the correct backend instead of the command being run again.

I think this is an awesome idea but out of scope for a first PR of introducing a `vllm` backend. I'll be sure to make a follow up issue for this feature.

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:70]: After doing some more research I'm removing the backend specific prefixes from the backend specific flags. The way we're using `click` for subcommands, flags between backends cannot be shared.

So running `ilab model serve vllm --help` would yield a different message than `ilab model serve --help` and `ilab model serve llama-cpp --help`. 

For this reason I'm not worried about users mixing up arguments as long as there are subcommands.

This is another reason why I believe going subcommands instead of having a  `--backend` flag to choose the backend.

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:65]: This is meant to be `--model` since we serve models not model paths!

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:25]: I meant to have this be the flags for either backend, not another subcommand.

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md:17]: @leseb Of the arguments you mentioned the one that stuck out to me the most is stability, in the sense that the next subcommand after `ilab model serve` has to be a backend. I understand wanting the flexibility for that subcommand in the future, but I'd argue we see the need for it now so we should just do it. 

Also I don't think it's the end of the world if the possibilities for the next subcommand aren't consistent. In fact if you start at the `ilab model` level, the only commonality is that the next level of sub-commands are related to models. The same principle applies to `ilab model serve`; the subcommands are related to model serving.

Comment (leseb) [reply docs/cli/ilab-model-serve.md:17]: > @leseb Of the arguments you mentioned the one that stuck out to me the most is stability, in the sense that the next subcommand after `ilab model serve` has to be a backend. I understand wanting the flexibility for that subcommand in the future, but I'd argue we see the need for it now so we should just do it.

Honestly, this isn't very future-proof. Since we're starting from scratch, we have the opportunity to design with future growth in mind and enhance user experience clarity.

> 
> Also I don't think it's the end of the world if the possibilities for the next subcommand aren't consistent. In fact if you start at the `ilab model` level, the only commonality is that the next level of sub-commands are related to models. The same principle applies to `ilab model serve`; the subcommands are related to model serving.

Yes, but today we think about `ilab model serve backend.` What if we later realized that the backend was just one of the possible sub-commands? It's not the end of the world, but it will degrade the user experience.

**Review by leseb — CHANGES_REQUESTED** [review]

**Review by markstur — COMMENTED** [review]
comments inline

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md]: This is a mistake and it should be updated in the most recent commit.

File: docs/cli/ilab-model-serve.md
Comment (alimaredia) [reply docs/cli/ilab-model-serve.md]: This was is a mistake that got cleaned up in the most recent commit. There are no flag prefixes anymore per the design doc

File: docs/cli/ilab-model-serve.md
Comment (russellb) [inline docs/cli/ilab-model-serve.md:5]: The CLI redesign link can be updated to point to the merged doc

File: docs/cli/ilab-model-serve.md
Comment (russellb) [inline docs/cli/ilab-model-serve.md:7]:
```suggestion
Specifically, this doc addresses the design of subcommands of `ilab model serve` that apply to different serving backends.
```

File: docs/cli/ilab-model-serve.md
Comment (russellb) [inline docs/cli/ilab-model-serve.md:59]: What if no backend is specified in the config?

how would express this config option? Config options and flags are the same thing. A config item here implies “—backend <foo>”

File: docs/cli/ilab-model-serve.md
Comment (russellb) [inline docs/cli/ilab-model-serve.md:87]: It’s probably easier to just have a pass through mechanism. I don’t see much value in replicating their config options in our own.

Comment (russellb) [reply docs/cli/ilab-model-serve.md:87]: I see the next line proposes that already… I’d prefer stating that pass through is the default answer unless we have a really good reason otherwise

**Review by russellb — COMMENTED** [review]
I left some minor comments. 

It seems the only thing blocking this is whether backend is a positional argument or a flag. 

If it was clear we’d have a bunch of backend specific options, I’d lean toward a positional argument to explicitly separate the set of options for each backend. 

It seems the doc proposes a generic pass through “—vllm-args” so it seems like differences will be minimal and clear enough. 

If I had to pick I’d go with the flag but I’m supportive of either one. We just need to resolve this quickly so implementation can proceed.

File: docs/cli/ilab-model-serve.md
Comment (n1hility) [reply docs/cli/ilab-model-serve.md:17]: While I was originally neutral, and perhaps leaning towards the subcommand. The recent proposals on the introduction of profiles shifted my perspective quite a bit on this one. 

I argue that with this capability users won't need to (or want to) care which backend is used to serve their model, all they really care about is that there model get served in a way that works and is optimal for the system they are using, and in a way that leads to good ilab results. Even if they do change it, often they will have only one workable choice for their system and use-case. As an example, running the full hifi SDG won't be possible with llama.cpp (at least today), and likewise, vllm is a bit too resource hungry for smaller scale systems.

Based on the above I think we should view this selection as an advanced option / parameter as opposed to a routine subcommand. There's also a compatibility aspect where users get guided into specifying a particular backend (since it's prominent when a subcommand) they might think they need to do this, and we could start to see more scripts and automation tied to it, which would make it harder for us to swap in a different technology we might introduce in the future (e.g say a server optimized for a particular model type, or specific GPU type). Granted this is still a risk with having the option at all, just likely less as an advanced option. 

Finally, I think there's an opportunity where we can make the technology more accessible by presenting higher level, more generic options. For those that are very specific to a particular backend we probably wouldn't want to continually sync ever low-level option, and those could be primarily pass through via the `[--vllm|--llama-cpp-args]` mechanism in this proposal.

Thoughts? 

I know we want to get this in asap, and ultimately I think we can make both approaches work. But I thought i would give this one more check in light of other recent design decisions.

**Review by n1hility — CHANGES_REQUESTED** [review]
Temp requesting changes just to verify the backend vs subcommand discussion

File: docs/cli/ilab-model-serve.md
Comment (n1hility) [reply docs/cli/ilab-model-serve.md:87]: +1 thats where I land. Ideally generic shared options that are intuitive and then passthrough so we don't have to keep a mirror in sync.

File: docs/cli/ilab-model-serve.md
Comment (n1hility) [reply docs/cli/ilab-model-serve.md:59]: This another reason  I think we should take another look at just making this an option (consistency between config and flags)

```suggestion
- If the model at `<PATH>` can be run by either backend, then a default backend defined in the config will be used as the model serving backend. In the case where there is ambiguity and a setting is not defined, a hardcoded preference will be used (all currently supported providers do not have this issue). A future profile specification will likely replace the hardcoded fallback. 
```

File: docs/cli/ilab-model-serve-backend.md
Comment (russellb) [inline docs/cli/ilab-model-serve-backend.md:58]: how about `--backend-arg` using click's `multiple=True` ? That way the semicolon-based parsing is not necessary. 

https://click.palletsprojects.com/en/7.x/options/#multiple-options

You will still need a way to separate arguments. For example `--backend-flag "foo 123"` makes sense, but we need to split that into two arguments. Splitting on spaces will break arguments that contain spaces.  Splitting on an `=` could work ... but any option could conflict with stuff that should be passed through unmangled.

The simplest implementation would be to always require `--backend-flag`. 

```
ilab model serve --backend vllm --backend-flag foo --backend-flag 123
```

... but that seems very tedious.

Another click option to consider is multi-value options https://click.palletsprojects.com/en/7.x/options/#multi-value-options

You could have `--backend-arg` (takes one arg) and `--backend-arg-opt` (takes 2 args).

```
ilab model serve --backend vllm --backend-arg boolean-arg --backend-arg arg-with-value 123`
```

**Review by russellb — DISMISSED** [review]
I'm good with the direction in general, though I have some comments on the syntax of the `--backend-args` option.

File: docs/cli/ilab-model-serve-backend.md
Comment (leseb) [reply docs/cli/ilab-model-serve-backend.md:58]: Thanks for raising this, IMO that's an implementation detail and I hope it's not blocking the design. From my perspective we need on the design to agree on the --backend-args approach, then how we implement it can be discussed during code reviews.
But I guess it's my fault and I went on with too many details :).

What do you think?

File: docs/cli/ilab-model-serve-backend.md
Comment (leseb) [reply docs/cli/ilab-model-serve-backend.md]: Gone.

File: docs/cli/ilab-model-serve-backend.md
Comment (russellb) [reply docs/cli/ilab-model-serve-backend.md:56]: > But I guess it's my fault and I went on with too many details :).

Right - I'm commenting based on the level of detail specified in the doc.

If you'd like to defer it as an implementation detail, you could remove a bunch of text and just say "we will provide a way to specify arguments to pass through to the backend" or something like that.

I think it's reasonable to cover it here, especially since the top issue debated here has been the UX, and this definitely UX impacting.

**Comment by alimaredia** [comment]
@leseb Thanks for moving this forward., I think `--backend-args` is very simple.

One question I have is do we need to be able to provide the ability to turn all of the knobs in the CLI that we could in a config file? AKA does every parameter need the ability to be set in the config and on the CLI?

I'm guessing the answer is yes, but if that's not that case why don't we just have:
1. Backend agnostic flags: ex `--model`
2. A flag to specify the backend `--backend`
3. A flag to specify the config file `--config`. We have docs for all of the config file options, and populate all values in the `serve` section via the config we generate with `ilab config init`.

At least with this approach there would be one source of truth, the implementation would be clean and simple, and we would not be passing and parsing flags within flags. I've never seen another project with flags similar to my original `--vllm-args` flag, or as the all encompassing `--backend-args` flag. I know it's simple, but my initial reaction is that it seems strange to pass flags within flags.

**Comment by mairin** [comment]
I would say yes for ease of use if you want to override one parameter or
whatnot in the config file.

**Comment by alimaredia** [comment]
The other thing I would say, is just passing the config file might make the world simpler if we we do default config setting with profiles like @cdoern has proposed. The user would only have to focus their attention on the config file. Charlie if you have any thoughts, please chime in.

File: docs/cli/ilab-model-serve-backend.md
Comment (n1hility) [reply docs/cli/ilab-model-serve-backend.md:56]: +1 ideally we want the design docs to be future referable (at least in the immediate term) to resolve uncertainties in the implementation. It's also ok o specify possible implementation options to help  explain a design as long as you are very clear it's a possible approach to be decided on later, and usually better if you list multiple options. For example, "--backend-args will support specifying one ore more options, possibly using  character delimited values or some other common multi-arg convention. The exact mechanism will be defined in a future proposal."

**Comment by leseb** [comment]
@alimaredia @mairin @n1hility @russellb, I'd like to hold off on discussing `--backend-args` until we have more people present. However, it seems we all agree on introducing a new `--backend` option. Can I get a confirmation on this? All my progress is stuck based on this single decision, and I have several PRs to submit. I'd like to make headway before I go away next week.

Much appreciated.

**Review by n1hility — APPROVED** [review]

**Comment by mairin** [comment]
I won't hold anything up but I don't agree with making --backend a flag. I have read back through comments here several times and do not see a rationale for this change. I think it is important to understand why we are going from subcommand to flag, as the subcommand approach was discussed extensively in brainstorming this devdoc and reviewed verbally with multiple folks - even rhtdan felt it was the way to go. I would feel better if I understood *why* we are going to a flag from a subcommand.

My core concern is from the user experience end. The user experience for this project is incredibly important and if we can have a delightful one it will be a differentiator. However, I am not considering other things like what limitations click might have. 

I think having --backendargs as a catch all and starting with that for the implemetation is a great idea

Again, I will not hold this one up bc we need to complete this and have smtg working and that is the most important.

**Comment by leseb** [comment]
Since some of the comments are not visible anymore after a push, I'd like to summarize why using a `ilab model serve --backend` flag is more appropriate than a positional argument like `ilab model serve <backend>`:

* Flexibility: Adding new backends or other options becomes more intuitive and consistent when using flags. It avoids the rigidity of positional arguments, making future expansions or modifications seamless. The flag `--backend` could be ported to other commands later (chat? generate) if we determine it makes sense.

* Consistency: Following a common pattern in CLI design where options are specified with flags ensures uniformity. This approach is clearer and more extendable, aligning with widely accepted CLI design principles.

* Clarity: Using `--backend` makes it immediately apparent that this is an option, enhancing the self-explanatory nature of the command. Users can quickly understand the command's structure and purpose. Also "Flags" signify optional parameters, aligning with the backend's nature as an auto-detected feature. Positional arguments typically indicate **required** inputs, which could confuse users about the backend's necessity. So the `--backend` flag help delineate optional from required parameters more effectively.

* Behavioral Context: Optional parameters, such as --backend, alter the program's behavior without being mandatory. This distinction is better communicated through flags, enhancing user comprehension.

As a side note: I'd argue that the only parameter that should be positional is the model path, as it's essential.

I'd to reference the guide from [CLIG.dev](https://clig.dev/#arguments-and-flags):

>Prefer flags to args. It’s a bit more typing, but it makes it much clearer what is going on. It also makes it easier to make changes to how you accept input in the future. Sometimes when using args, it’s impossible to add new input without breaking existing behavior or creating ambiguity.

In conclusion, using `--backend` for the backend ensures flexibility, consistency, clarity, better integration, and aligns with best practices for CLI design.

Thanks!

@mairin I hope this will change your mind :). Are you keeping your request change to express your disagreement with the latest proposal? I'm seeking clarification because your last comment said you won't hold this up. I'm fine if you want to keep it and have someone else merge if this gets other approvals.

**Comment by mairin** [comment]
Sébastien and I had a call this morning and I am on board with this design now. The rationale had been squashed and he posted it above, and I also misunderstood part of how the flag was proposed to work. I am on board with this now. My main concern was making sure users do not try to use on set of arguments they learned to use with llamacpp and try to xfer them to a vllm or other backend and have a poor experience. Some things we discussed is capturing when that happens and erroring out gracefully, or letting the user know it won't work if we can detect that's going on (with an out to override and try it anyway), and also giving suggesting based on what the argument for backend A is with the equivalent argument for backend B.

**Review by mairin — APPROVED** [review]

**Pull request merged**
